File: RDMPEG/RDMPEGConverter/RDMobileFFmpegOperation.m
//
//  RDMobileFFmpegOperation.m
//  RDMPEG
//
//  Created by Artem on 24.08.2020.
//  Copyright © 2020 Readdle. All rights reserved.
//

#import "RDMobileFFmpegOperation.h"
#import "RDMPEGOperation+Protected.h"
#import <FFmpegKit/FFmpegKit.h>
#import <Log4Cocoa/Log4Cocoa.h>

#import <RDMPEG/RDMPEG-Swift.h>


@interface RDMobileFFmpegOperation()

@property (nonatomic,strong)NSArray<NSString *> *arguments;
@property (nonatomic,copy)RDMobileFFmpegOperationResultBlock resultBlock;
@property (nonatomic,copy)RDMobileFFmpegOperationStatisticsBlock statisticsBlock;
@property (nonatomic, copy, nullable) RDMobileFFmpegOperationLogBlock logBlock;
@property (atomic, strong) FFmpegSession *session;

@end


@implementation RDMobileFFmpegOperation

- (instancetype)initWithArguments:(NSArray<NSString *> *)arguments
                  statisticsBlock:(RDMobileFFmpegOperationStatisticsBlock)statisticsBlock
                      resultBlock:(RDMobileFFmpegOperationResultBlock)resultBlock{
    return [self initWithArguments:arguments statisticsBlock:statisticsBlock resultBlock:resultBlock logBlock:nil];
}

- (instancetype)initWithArguments:(NSArray<NSString *> *)arguments
                  statisticsBlock:(RDMobileFFmpegOperationStatisticsBlock)statisticsBlock
                      resultBlock:(RDMobileFFmpegOperationResultBlock)resultBlock
                         logBlock:(__nullable RDMobileFFmpegOperationLogBlock)logBlock {
    NSParameterAssert(arguments);
    if(arguments == nil){
        return nil;
    }
    self = [super init];
    if(self){
        self.arguments = arguments;
        self.resultBlock = resultBlock;
        self.statisticsBlock = statisticsBlock;
        self.logBlock = logBlock;
    }
    return self;
}

- (void)main {
    NSParameterAssert(self.session == nil);
    
    __weak __typeof(self) const weakSelf = self;
    
    self.session =
    [FFmpegKit
     executeWithArgumentsAsync:self.arguments
     withExecuteCallback:^(id<Session> session) {
        if (weakSelf.resultBlock) {
            weakSelf.resultBlock([[session getReturnCode] getValue]);
        }
        
        [weakSelf completeOperation];
     }
     withLogCallback:^(Log *log) {
        if (weakSelf.logBlock) {
            weakSelf.logBlock([log getMessage], [log getLevel]);
        }
        log4CDebug(@"level: %@, message: %@", @([log getLevel]), [log getMessage]);
     }
     withStatisticsCallback:^(Statistics *statistics) {
        if (weakSelf.statisticsBlock) {
            weakSelf.statisticsBlock([[RDMobileFFmpegStatistics alloc] initWithStatistics:statistics]);
        }
    }];
}

- (void)cancel{
    [super cancel];
    
    long sessionId = [self.session getSessionId];
    [FFmpegKit cancel:sessionId];
}

+ (BOOL)isReturnCodeCancel:(int)code{
    return code == ReturnCodeCancel;
}

+ (BOOL)isReturnCodeSuccess:(int)code{
    return code == ReturnCodeSuccess;
}

@end

File: RDMPEG/RDMPEGConverter/RDMobileFFmpegOperation.swift
//
//  RDMobileFFmpegOperation.swift
//  RDMPEG
//
//  Created by Max Berezhnoy on 16/07/2024.
//  Copyright © 2024 Readdle. All rights reserved.
//

import Foundation
import ffmpegkit
import Log4Cocoa

@objcMembers
public class RDMobileFFmpegOperation: RDMPEGOperation, @unchecked Sendable {

    public typealias StatisticsBlock = (RDMobileFFmpegStatistics) -> Void
    public typealias ResultBlock = (Int32) -> Void
    public typealias LogBlock = (String, Int32) -> Void

    private let arguments: [String]
    private let resultBlock: ResultBlock
    private let statisticsBlock: StatisticsBlock
    private let logBlock: LogBlock?
    private var session: FFmpegSession?

    public init(arguments: [String], statisticsBlock: @escaping StatisticsBlock, resultBlock: @escaping ResultBlock) {
        self.arguments = arguments
        self.statisticsBlock = statisticsBlock
        self.resultBlock = resultBlock
        self.logBlock = nil
        super.init()
    }

    public init(
        arguments: [String],
        statisticsBlock: @escaping StatisticsBlock,
        resultBlock: @escaping ResultBlock,
        logBlock: LogBlock?
    ) {
        self.arguments = arguments
        self.statisticsBlock = statisticsBlock
        self.resultBlock = resultBlock
        self.logBlock = logBlock
        super.init()
    }

    override public func main() {
        assert(session == nil)

        session = FFmpegKit.execute(
            withArgumentsAsync: arguments,
            withExecuteCallback: { [weak self] session in
                guard let self, let session else { return }

                self.resultBlock(session.getReturnCode().getValue())
                self.completeOperation()
            },
            withLogCallback: { [weak self] log in
                guard let self, let log else { return }

                if let logBlock = self.logBlock {
                    logBlock(log.getMessage() ?? "", log.getLevel())
                }

                log4Debug("level: \(log.getLevel()), message: \(log.getMessage() ?? "")")
            },
            withStatisticsCallback: { [weak self] statistics in
                guard let self, let statistics else { return }

                self.statisticsBlock(RDMobileFFmpegStatistics(statistics: statistics))
            }
        )
    }

    override public func cancel() {
        super.cancel()

        if let sessionId = session?.getId() {
            FFmpegKit.cancel(sessionId)
        }
    }

    public class func isReturnCodeCancel(_ code: Int32) -> Bool {
        return code == ReturnCodeEnum.cancel.rawValue
    }

    public class func isReturnCodeSuccess(_ code: Int32) -> Bool {
        return code == ReturnCodeEnum.success.rawValue
    }
}
